### PR TITLE
Add support for clair notifier

### PR DESCRIPTION
**Pull request by SuperSandro2000**
Depends on #194

### CODE REVIEW

File: internal/tasks/manifests_test.go [hunk 1]
```diff
@@ -753,3 +753,69 @@ func TestCheckVulnerabilitiesForNextManifestWithError(t *testing.T) {
...
+
+	//also setup an image list manifest containing those images (so that we have
+	//some manifest-manifest refs to play with)
+	imageList := test.GenerateImageList(images[0], images[1])
+	imageList.MustUpload(t, s, fooRepoRef, "")
+
+	tr, tr0 := easypg.NewTracker(t, s.DB.DbMap.Db)
+	tr0.AssertEqualToFile("fixtures/notifier-check-setup.sql")
+
+	//setup our Clair API double (TODO use test.WithClairDouble instead)
```
Comment (SuperSandro2000) [inline internal/tasks/manifests_test.go:777]: Rebase after other branch that touches this todo is merged

**Comment by SuperSandro2000** [comment]
@majewsky PTAL

File: internal/api/clairintegration/api.go [hunk 2]
```diff
@@ -71,3 +95,95 @@ func (a *API) reverseProxyToClair(w http.ResponseWriter, r *http.Request) {
...
+			for _, entry := range value {
+				if entry == c.NotificationSecret {
+					hasSecretHeader = true
+					break auth
+				}
+			}
+		}
+	}
+
+	if !hasSecretHeader {
```
Comment (majewsky) [inline internal/api/clairintegration/api.go:119]:
```suggestion
secretHeader := r.Header[http.CanonicalHeaderKey("X-KEPPEL-CLAIR_NOTIFICATION-SECRET")]
if !slices.Contains(secretHeader, c.NotificationSecret) {
```

File: internal/api/clairintegration/api.go [hunk 3]
```diff
@@ -71,3 +95,95 @@ func (a *API) reverseProxyToClair(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "unauthorized", http.StatusUnauthorized)
+		return
+	}
+
+	reqBodyBytes, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnprocessableEntity)
+		logg.Error(err.Error())
+		return
+	}
```
Comment (majewsky) [inline internal/api/clairintegration/api.go:129]: IO errors should be a 5xx in my opinion.
```suggestion
if respondwith.ErrorText(w, err) {
  return
}
```

**Review by majewsky — COMMENTED** [review]
First round of comments.

File: internal/clair/client.go [hunk 4]
```diff
@@ -40,15 +44,21 @@ type Client struct {
...
+func (c *Client) RequestURL(pathElements ...string) string {
 	requestURL := c.BaseURL
 	requestURL.Path = path.Join(c.BaseURL.Path, path.Join(pathElements...))
 	return requestURL.String()
 }
 
-func (c *Client) doRequest(req *http.Request, respBody interface{}) error {
+func (c *Client) DoRequest(req *http.Request, respBody interface{}) error {
+	if respBody != nil && reflect.ValueOf(respBody).Kind() != reflect.Pointer {
+		logg.Fatal("DoRequest only takes nil or a pointer as a second argument!")
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply internal/clair/client.go:59]: :(

```
method must have no type parameters syntax
```

File: internal/api/clairintegration/api.go (hunk 3 above)
Comment (majewsky) [reply internal/api/clairintegration/api.go:116]: This was marked as resolved without any change.

File: internal/clair/index.go [hunk 5]
```diff
@@ -53,23 +54,29 @@ type ManifestState struct {
...
 
+// Based on upstream type with the same name https://github.com/quay/claircore/blob/main/indexreport.go
 type indexReport struct {
-	Digest       string `json:"manifest_hash"`
-	State        string `json:"state"`
-	ErrorMessage string `json:"err"`
-	//there are more fields, but we are not interested in them
+	Digest string `json:"manifest_hash"`
+	State  string `json:"state"`
+	Err    string `json:"err"`
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply internal/clair/index.go:61]: I think I was shortly trying to use the upstream type

File: internal/api/clairintegration/api.go [hunk 6]
```diff
@@ -71,3 +95,87 @@ func (a *API) reverseProxyToClair(w http.ResponseWriter, r *http.Request) {
...
+	// collect all notifications
+	var notifications []clair.Notification
+	notificationResponse, err := c.GetNotification(callbackPath, "")
+	if err != nil {
+		return err
+	}
+	notifications = append(notifications, notificationResponse.Notifications...)
+
+	for notificationResponse.Page.Next != nil && notificationResponse.Page.Next.String() != "" {
+		notificationResponse, err = c.GetNotification(callbackPath, "")
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply internal/api/clairintegration/api.go:187]: Yes. The notifications themselves only contain a uuid and the notification contents, not the full callback URL.

https://quay.github.io/clair/reference/api.html?search=state#:~:text=to%20continue%20paging-,PagedNotifications,-%7B%0A%20%20%22page%22

[the remark this answers is not in the document]
Comment (majewsky) [reply internal/api/clairintegration/api.go:187]: My question is if it's correct that the `callbackPath` does not change between loop iterations here.

**Pull request closed without merging**